# The 2018 Final Four games are played under a full moon. When was the last time there was an upset in a Final Four game under a full moon?

## Step 1

### Let's start by taking a look at a dataset of the phases of the moon. The moon has [4 principal phases](https://en.wikipedia.org/wiki/Lunar_phase#Phases_of_the_Moon): the new moon, the first quarter, the full moon, and the last quarter. A full moon occurs when Earth is located directly between the sun and the moon, causing the part of the moon closest to Earth to appear fully sunlit, while the far side is completely dark.

In [0]:
from pandas.io import gbq
project_id = '[YOUR_PROJECT_ID]'

In [14]:
moons_q = """
SELECT phase, phase_emoji, peak_datetime
FROM `bigquery-public-data.moon_phases.moon_phases`
ORDER BY peak_datetime ASC
"""

moons = gbq.read_gbq(query=moons_q, dialect ='standard', project_id=project_id)
moons.head(25)

Requesting query... ok.
Job ID: job_NMtV3iD9oiRzJ_XQaRQJ-_keIg5O
Query running...
Query done.
Cache hit.

Retrieving results...
Got 19789 rows.

Total time taken 2.56 s.
Finished at 2018-03-05 07:33:15.


phase phase_emoji        peak_datetime
0       Full Moon           🌕  1700-01-05T10:30:00
1    Last Quarter           🌗  1700-01-12T03:34:00
2        New Moon           🌑  1700-01-20T04:20:00
3   First Quarter           🌓  1700-01-28T05:13:00
4       Full Moon           🌕  1700-02-03T21:05:00
5    Last Quarter           🌗  1700-02-10T18:59:00
6        New Moon           🌑  1700-02-18T23:33:00
7   First Quarter           🌓  1700-02-26T16:37:00
8       Full Moon           🌕  1700-03-05T07:38:00
9    Last Quarter           🌗  1700-03-12T12:29:00
10       New Moon           🌑  1700-03-20T16:47:00
11  First Quarter           🌓  1700-03-28T00:42:00
12      Full Moon           🌕  1700-04-03T18:14:00
13   Last Quarter           🌗  1700-04-11T07:07:00
14       New Moon           🌑  1700-04-19T06:51:00
15  First Quarter           🌓  1700-04-26T06:41:00
16      Full Moon           🌕  1700-05-03T05:11:00
17   Last Quarter           🌗  1700-05-11T01:45:00
18       New Moon           🌑  1700-05-18T17:45:00
19  First Quarter           🌓  1700-05-25T11:53:00
20      Full Moon           🌕  1700-06-01T16:58:00
21   Last Quarter           🌗  1700-06-09T19:10:00
22       New Moon           🌑  1700-06-17T02:16:00
23  First Quarter           🌓  1700-06-23T17:35:00
24      Full Moon           🌕  1700-07-01T06:08:00

## Step 2

### These timestamps are in [UTC time](https://en.wikipedia.org/wiki/Coordinated_Universal_Time), so we'll need to convert these to a different timezone if we want to find the date of the full moon in the United States:

In [15]:
moons_et_q = """
SELECT 
  phase,
  phase_emoji,
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") phase_peak_nyc
FROM `bigquery-public-data.moon_phases.moon_phases`
ORDER BY peak_datetime ASC
"""

moons_et = gbq.read_gbq(query=moons_et_q, dialect ='standard', project_id=project_id)
moons_et.head(25)

Requesting query... ok.
Job ID: job_7UlI_-67yriq6C4OZcXxIplbRik4
Query running...
Query done.
Cache hit.

Retrieving results...
Got 19789 rows.

Total time taken 2.13 s.
Finished at 2018-03-05 07:33:23.


phase phase_emoji       phase_peak_nyc
0       Full Moon           🌕  1700-01-05T05:33:58
1    Last Quarter           🌗  1700-01-11T22:37:58
2        New Moon           🌑  1700-01-19T23:23:58
3   First Quarter           🌓  1700-01-28T00:16:58
4       Full Moon           🌕  1700-02-03T16:08:58
5    Last Quarter           🌗  1700-02-10T14:02:58
6        New Moon           🌑  1700-02-18T18:36:58
7   First Quarter           🌓  1700-02-26T11:40:58
8       Full Moon           🌕  1700-03-05T02:41:58
9    Last Quarter           🌗  1700-03-12T07:32:58
10       New Moon           🌑  1700-03-20T11:50:58
11  First Quarter           🌓  1700-03-27T19:45:58
12      Full Moon           🌕  1700-04-03T13:17:58
13   Last Quarter           🌗  1700-04-11T02:10:58
14       New Moon           🌑  1700-04-19T01:54:58
15  First Quarter           🌓  1700-04-26T01:44:58
16      Full Moon           🌕  1700-05-03T00:14:58
17   Last Quarter           🌗  1700-05-10T20:48:58
18       New Moon           🌑  1700-05-18T12:48:58
19  First Quarter           🌓  1700-05-25T06:56:58
20      Full Moon           🌕  1700-06-01T12:01:58
21   Last Quarter           🌗  1700-06-09T14:13:58
22       New Moon           🌑  1700-06-16T21:19:58
23  First Quarter           🌓  1700-06-23T12:38:58
24      Full Moon           🌕  1700-07-01T01:11:58

## Step 3

### Since the exact moment of the full moon peak may or may not occur “at night” in whatever time zone you happen to be (or, for our purposes, wherever basketball games are being played), we need to come up with a reasonable window of time around the full moon peak in which a game might have transpired. 

### NCAA Final Four games tend to happen between 6 PM and 12 PM EST, so let's start with that as our baseline game window. We'll then add an 18-hour cushion on either side of the game time to account for the passage of any moon phase peak. This establishes a 42-hour window to capture tournament games near or during peak phases of the moon, and excludes games that occurred outside of those 42-hour windows.

### Now that our window is defined, we'll use an *inner* join so that we only see games that match with our moon phase windows. Here are the games that were played within 18 hours of any moon phase peak:

In [13]:
games_all_phases_q = """
SELECT
  game_date,
  round, 
  moon.phase,
  moon.phase_emoji,
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") phase_peak_nyc,
  win_seed, 
  win_market, 
  lose_seed, 
  lose_market
FROM `bigquery-public-data.ncaa_basketball.mbb_historical_tournament_games` as game
INNER JOIN `bigquery-public-data.moon_phases.moon_phases` as moon on 
  (DATETIME(TIMESTAMP(peak_datetime), "America/New_York") > DATETIME(game_date) AND  # <<<<< this line is the lower limit, restricting the results to just games where the phase peak occurs within 18 hours of the game start time.
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") < DATETIME_ADD(DATETIME(game_date), INTERVAL 42 HOUR))# <<<<< this line is the upper limit, restricting the results to just games where the phase peak occurs within 18 hours of the game end time.
ORDER BY game_date DESC
"""

games_all_phases = gbq.read_gbq(query=games_all_phases_q, dialect ='standard', project_id=project_id)
games_all_phases.head(25)

Requesting query... ok.
Job ID: job_r-e7ZqqeFC2YlIgn34yo00-0T0wC
Query running...
Query done.
Cache hit.

Retrieving results...
Got 435 rows.

Total time taken 1.05 s.
Finished at 2018-03-05 07:32:50.


game_date  round          phase phase_emoji       phase_peak_nyc  \
0   2017-04-03      2  First Quarter           🌓  2017-04-03T14:39:00   
1   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
2   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
3   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
4   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
5   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
6   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
7   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
8   2017-03-19     32   Last Quarter           🌗  2017-03-20T11:58:00   
9   2016-03-15     68  First Quarter           🌓  2016-03-15T13:03:00   
10  2016-03-15     68  First Quarter           🌓  2016-03-15T13:03:00   
11  2015-04-04      4      Full Moon           🌕  2015-04-04T08:05:00   
12  2015-04-04      4      Full Moon           🌕  2015-04-04T08:05:00   
13  2015-03-27     16  First Quarter           🌓  2015-03-27T03:43:00   
14  2015-03-27     16  First Quarter           🌓  2015-03-27T03:43:00   
15  2015-03-27     16  First Quarter           🌓  2015-03-27T03:43:00   
16  2015-03-27     16  First Quarter           🌓  2015-03-27T03:43:00   
17  2015-03-26     16  First Quarter           🌓  2015-03-27T03:43:00   
18  2015-03-26     16  First Quarter           🌓  2015-03-27T03:43:00   
19  2015-03-26     16  First Quarter           🌓  2015-03-27T03:43:00   
20  2015-03-26     16  First Quarter           🌓  2015-03-27T03:43:00   
21  2015-03-20     64       New Moon           🌑  2015-03-20T05:36:00   
22  2015-03-20     64       New Moon           🌑  2015-03-20T05:36:00   
23  2015-03-20     64       New Moon           🌑  2015-03-20T05:36:00   
24  2015-03-20     64       New Moon           🌑  2015-03-20T05:36:00   

   win_seed          win_market lose_seed           lose_market  
0        01      North Carolina        01               Gonzaga  
1        07      South Carolina        02                  Duke  
2        01              Kansas        09        Michigan State  
3        01      North Carolina        08              Arkansas  
4        03              Oregon        11          Rhode Island  
5        03              Baylor        11                   USC  
6        02            Kentucky        10         Wichita State  
7        07            Michigan        02            Louisville  
8        03                UCLA        06            Cincinnati  
9        16  Florida Gulf Coast        16   Fairleigh Dickinson  
10       11       Wichita State        11            Vanderbilt  
11       01                Duke        07        Michigan State  
12       01           Wisconsin        01              Kentucky  
13       04          Louisville        08  North Carolina State  
14       07      Michigan State        03              Oklahoma  
15       01                Duke        05                  Utah  
16       02             Gonzaga        11                  UCLA  
17       01           Wisconsin        04        North Carolina  
18       02             Arizona        06                Xavier  
19       01            Kentucky        05         West Virginia  
20       03          Notre Dame        07         Wichita State  
21       01                Duke        16         Robert Morris  
22       05       Northern Iowa        12               Wyoming  
23       07      Michigan State        10               Georgia  
24       03            Oklahoma        14                Albany

## Step 4

### Now, let's limit it to only games on the full moon.

In [16]:
games_full_moons_q = """
SELECT
  game_date,
  round, 
  moon.phase,
  moon.phase_emoji,
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") phase_peak_nyc,
  win_seed, 
  win_market, 
  lose_seed, 
  lose_market
FROM `bigquery-public-data.ncaa_basketball.mbb_historical_tournament_games` as game
INNER JOIN `bigquery-public-data.moon_phases.moon_phases` as moon on 
  (DATETIME(TIMESTAMP(peak_datetime), "America/New_York") > DATETIME(game_date) AND  # <<<<< this line is the lower limit, restricting the results to just games where the phase peak occurs within 18 hours of the game start time.
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") < DATETIME_ADD(DATETIME(game_date), INTERVAL 42 HOUR))# <<<<< this line is the upper limit, restricting the results to just games where the phase peak occurs within 18 hours of the game end time.
WHERE moon.phase_emoji = "🌕" # <<<<<<<<<<<< this is the new line
ORDER BY game_date DESC
"""

games_full_moons = gbq.read_gbq(query=games_full_moons_q, dialect ='standard', project_id=project_id)
games_full_moons.head(25)

Requesting query... ok.
Job ID: job_NhYoUXABaoNKzb-23OrHS60exWp7
Query running...
Query done.
Processed: 608.7 KB
Standard price: $0.00 USD

Retrieving results...
Got 137 rows.

Total time taken 2.45 s.
Finished at 2018-03-05 07:36:54.


game_date  round      phase phase_emoji       phase_peak_nyc win_seed  \
0   2015-04-04      4  Full Moon           🌕  2015-04-04T08:05:00       01   
1   2015-04-04      4  Full Moon           🌕  2015-04-04T08:05:00       01   
2   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       03   
3   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       03   
4   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       04   
5   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       04   
6   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       08   
7   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       12   
8   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       02   
9   2011-03-19     32  Full Moon           🌕  2011-03-19T14:10:00       02   
10  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       11   
11  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       01   
12  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       08   
13  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       03   
14  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       11   
15  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       02   
16  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       01   
17  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       03   
18  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       01   
19  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       04   
20  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       08   
21  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       02   
22  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       09   
23  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       10   
24  2011-03-18     64  Full Moon           🌕  2011-03-19T14:10:00       05   

               win_market lose_seed        lose_market  
0                    Duke        07     Michigan State  
1               Wisconsin        01           Kentucky  
2             Connecticut        06         Cincinnati  
3                     BYU        11            Gonzaga  
4               Wisconsin        05       Kansas State  
5                Kentucky        05      West Virginia  
6                  Butler        01         Pittsburgh  
7                Richmond        13     Morehead State  
8                 Florida        07               UCLA  
9         San Diego State        07             Temple  
10  Virginia Commonwealth        06         Georgetown  
11                   Duke        16            Hampton  
12               Michigan        09          Tennessee  
13               Syracuse        14      Indiana State  
14              Marquette        06             Xavier  
15         North Carolina        15       LIU Brooklyn  
16                 Kansas        16  Boston University  
17                 Purdue        14        St. Peter's  
18             Ohio State        16               UTSA  
19                  Texas        13            Oakland  
20           George Mason        09          Villanova  
21             Notre Dame        15              Akron  
22               Illinois        08               UNLV  
23          Florida State        07          Texas A&M  
24                Arizona        12            Memphis

### Ok, now we're getting somewhere. Of the 2117 games played in the NCAA tournament since 1985, only 137 have been played under a full moon.

### 4A: As an aside, that's only slightly more than you'd expect. According to Wikipedia, a [Synodic lunar month](https://en.wikipedia.org/wiki/Lunar_month#Synodic_month) averages 29.530587981 days. We'd expect this many games under our full moon window:

In [17]:
(42.0/24) * (2117/29.530587981)


125.45466424114679

## Step 5

### Back to business... as we know, the Final Four in 2018 is on a full moon (the peak is at 8:37AM ET the morning of the games), so let's only look at Final Four games.

In [18]:
games_full_moons_4_q = """
SELECT
  game_date,
  round, 
  moon.phase,
  moon.phase_emoji,
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") phase_peak_nyc,
  win_seed, 
  win_market, 
  lose_seed, 
  lose_market
FROM `bigquery-public-data.ncaa_basketball.mbb_historical_tournament_games` as game
INNER JOIN `bigquery-public-data.moon_phases.moon_phases` as moon on 
  (DATETIME(TIMESTAMP(peak_datetime), "America/New_York") > DATETIME(game_date) AND  # <<<<< this line is the lower limit, restricting the results to just games where the phase peak occurs within 18 hours of the game start time.
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") < DATETIME_ADD(DATETIME(game_date), INTERVAL 42 HOUR))# <<<<< this line is the upper limit, restricting the results to just games where the phase peak occurs within 18 hours of the game end time.
WHERE moon.phase_emoji = "🌕"
AND round = 4 # <<<<<<<<<<<< this is the new line
ORDER BY game_date DESC
"""

games_full_moons_4 = gbq.read_gbq(query=games_full_moons_4_q, dialect ='standard', project_id=project_id)
games_full_moons_4.head()


Requesting query... ok.
Job ID: job_8PFaF07_ALxKE7DkLSZLZ522nHrC
Query running...
Query done.
Processed: 608.7 KB
Standard price: $0.00 USD

Retrieving results...
Got 6 rows.

Total time taken 2.39 s.
Finished at 2018-03-05 07:39:34.


game_date  round      phase phase_emoji       phase_peak_nyc win_seed  \
0  2015-04-04      4  Full Moon           🌕  2015-04-04T08:05:00       01   
1  2015-04-04      4  Full Moon           🌕  2015-04-04T08:05:00       01   
2  1991-03-30      4  Full Moon           🌕  1991-03-30T02:17:00       02   
3  1991-03-30      4  Full Moon           🌕  1991-03-30T02:17:00       03   
4  1988-04-02      4  Full Moon           🌕  1988-04-02T04:21:00       01   

  win_market lose_seed     lose_market  
0  Wisconsin        01        Kentucky  
1       Duke        07  Michigan State  
2       Duke        01            UNLV  
3     Kansas        01  North Carolina  
4   Oklahoma        01         Arizona

## Step 6

### It's only 6 games! It's easy to see, but let's run the query to only return the upsets. Since we've organized the tournament_games dataset to show the seed of the winning team and the seed of the losing team, it's easy to find them: they're the games where the winning seed is a larger number than the losing seed.

In [19]:
upsets_full_moons_4_q = """
SELECT
  game_date,
  round, 
  moon.phase,
  moon.phase_emoji,
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") phase_peak_nyc,
  win_seed, 
  win_market, 
  lose_seed, 
  lose_market
FROM `bigquery-public-data.ncaa_basketball.mbb_historical_tournament_games` as game
INNER JOIN `bigquery-public-data.moon_phases.moon_phases` as moon on 
  (DATETIME(TIMESTAMP(peak_datetime), "America/New_York") > DATETIME(game_date) AND  # <<<<< this line is the lower limit, restricting the results to just games where the phase peak occurs within 18 hours of the game start time.
  DATETIME(TIMESTAMP(peak_datetime), "America/New_York") < DATETIME_ADD(DATETIME(game_date), INTERVAL 42 HOUR))# <<<<< this line is the upper limit, restricting the results to just games where the phase peak occurs within 18 hours of the game end time.
WHERE moon.phase_emoji = "🌕"
AND round = 4
AND win_seed > lose_seed # <<<<<<<<<<<< this is the new line
ORDER BY game_date DESC
"""

upsets_full_moons_4 = gbq.read_gbq(query=upsets_full_moons_4_q, dialect ='standard', project_id=project_id)
upsets_full_moons_4.head()



Requesting query... ok.
Job ID: job_F9i3fuE7O9xREHRho4rdNM0IuPat
Query running...
Query done.
Processed: 608.7 KB
Standard price: $0.00 USD

Retrieving results...
Got 3 rows.

Total time taken 1.48 s.
Finished at 2018-03-05 07:40:40.


game_date  round      phase phase_emoji       phase_peak_nyc win_seed  \
0  1991-03-30      4  Full Moon           🌕  1991-03-30T02:17:00       03   
1  1991-03-30      4  Full Moon           🌕  1991-03-30T02:17:00       02   
2  1988-04-02      4  Full Moon           🌕  1988-04-02T04:21:00       06   

  win_market lose_seed     lose_market  
0     Kansas        01  North Carolina  
1       Duke        01            UNLV  
2     Kansas        02            Duke

# There have been three upsets in the Final Four under a full moon since 1985, two of which both happened in 1991!